csrank/objectranking/lambdamart.py

```diff
@@ -0,0 +1,370 @@
+                scores_as_list.append(relscore_train[x,y])
+        
+        #Check if both the dimensions are the same
+        assert(len(features_as_list)==len(scores_as_list))
+        
+        #convert to numpy and resize the arrays 
+        features = np.asarray(features_as_list)
+        scores_unflat = np.array(scores_as_list)
+        scores = np.reshape(scores_unflat,(len(scores_unflat),1))
+
+        #Concatenate the reshaped arrays and return as trainin data
+        train_data = np.concatenate((scores,features),axis=1)
+
+        return train_data
+
+    def _i_dcg(self,scores,k=None):
+        """
+            Returns the ideal DCG value for the given scores
+        """
+        scores = [score for score in sorted(scores)[::-1]]
+        if k == None:
+            return np.sum([(np.power(2, scores[i]) - 1) / np.log2(i + 2) for i in range(len(scores))])
+
+        elif k > 0:
+            return np.sum([(np.power(2, scores[i]) - 1) / np.log2(i + 2) for i in range(len(scores[:k]))])
+
+    def _group_queries(self, training_data, qid_index):
+        query_indexes = {}
+        index = 0
+        for record in training_data:
+            query_indexes.setdefault(record[qid_index], [])
+            query_indexes[record[qid_index]].append(index)
+            index += 1
+        return query_indexes
+
+    def fit(self, X, y, **kwargs):
+        """
+            Fit a LambdaMART algorithm to the provided X and y arrays where X contains the features and y being the relevance scores.
+
+            Parameters
+            ----------
+            X : numpy array
+                (n_instances, n_objects, n_features)
+                Feature vectors of the objects
+            Y : numpy array
+                (n_instances, n_objects)
+                Rankings of the given objects
+            **kwargs
+                Keyword arguments for the fit function
+
+        """
+
+        training_data = self._prepare_train_data(X, y)
+        predicted_scores = np.zeros(len(training_data))
+        query_indexes = self._group_queries(training_data, 1)
+        query_keys = query_indexes.keys()
+        true_scores = [training_data[query_indexes[query], 0] for query in query_keys]
+
+        sorted_query_pair = []
+        for query_scores in true_scores:
+            temp = sorted(query_scores, reverse=True)
+            pairs = []
+            for i in range(len(temp)):
+                for j in range(len(temp)):
+                    if temp[i] > temp[j]:
+                        pairs.append((i,j))
+            sorted_query_pair.append(pairs)
+
+        idcg = [self._i_dcg(scores) for scores in true_scores]
+
+        for k in range(self.number_of_trees):
+            print("Tree %d" % (k))
+            lambdas = np.zeros(len(predicted_scores))
+            w = np.zeros(len(predicted_scores))
+            pred_scores = [predicted_scores[query_indexes[query]] for query in query_keys]
+            
+            pool = Pool()
```

Comment:
If you use a pool for multiprocessing, make sure that the user can pass the number of processes it `fit`.
Also, use a `with` environment like so:
```python
with Pool(5) as p:
    print(p.map(f, [1, 2, 3]))
```

Maybe even use joblib.